core/src/ledger/masp_utils.rs

```diff
@@ -71,42 +62,27 @@ pub fn update_note_commitment_tree(
 /// Handle a MASP transaction.
 pub fn handle_masp_tx(
     ctx: &mut (impl StorageRead + StorageWrite),
-    transfer: &Transfer,
     shielded: &Transaction,
+    pin_key: Option<&str>,
 ) -> Result<()> {
-    let masp_addr = MASP;
-    let head_tx_key = Key::from(masp_addr.to_db_key())
-        .push(&HEAD_TX_KEY.to_owned())
-        .expect("Cannot obtain a storage key");
-    let current_tx_idx: u64 =
-        ctx.read(&head_tx_key).unwrap_or(None).unwrap_or(0);
-    let current_tx_key = Key::from(masp_addr.to_db_key())
-        .push(&(TX_KEY_PREFIX.to_owned() + &current_tx_idx.to_string()))
-        .expect("Cannot obtain a storage key");
-    // Save the Transfer object and its location within the blockchain
-    // so that clients do not have to separately look these
-    // up
-    let record: (Epoch, BlockHeight, TxIndex, Transfer, Transaction) = (
-        ctx.get_block_epoch()?,
-        ctx.get_block_height()?,
-        ctx.get_tx_index()?,
-        transfer.clone(),
-        shielded.clone(),
-    );
-    ctx.write(&current_tx_key, record)?;
-    ctx.write(&head_tx_key, current_tx_idx + 1)?;
     // TODO: temporarily disabled because of the node aggregation issue in WASM.
     // Using the host env tx_update_masp_note_commitment_tree or directly the
     // update_note_commitment_tree function as a  workaround instead
     // update_note_commitment_tree(ctx, shielded)?;
     reveal_nullifiers(ctx, shielded)?;
 
     // If storage key has been supplied, then pin this transaction to it
-    if let Some(key) = &transfer.key {
-        let pin_key = Key::from(masp_addr.to_db_key())
+    if let Some(key) = pin_key {
```

Comment:
you wrote all these nice factory functions for the other keys, why not this one?